br/pkg/storage/s3.go

```diff
@@ -528,28 +536,47 @@ func (rs *S3Storage) open(
 	path string,
 	startOffset, endOffset int64,
 ) (io.ReadCloser, RangeInfo, error) {
+	var err error
 	input := &s3.GetObjectInput{
 		Bucket: aws.String(rs.options.Bucket),
 		Key:    aws.String(rs.options.Prefix + path),
 	}
 
 	// always set rangeOffset to fetch file size info
 	// s3 endOffset is inclusive
+	isFullRangeRequest := false
 	var rangeOffset *string
-	if endOffset > startOffset {
+	switch {
+	case endOffset > startOffset:
 		rangeOffset = aws.String(fmt.Sprintf("bytes=%d-%d", startOffset, endOffset-1))
-	} else {
+	case startOffset == 0:
+		isFullRangeRequest = true
+	default:
 		rangeOffset = aws.String(fmt.Sprintf("bytes=%d-", startOffset))
 	}
 	input.Range = rangeOffset
 	result, err := rs.svc.GetObjectWithContext(ctx, input)
+	log.L().Debug("S3 get object", zap.Any("request", input), zap.Any("response", result))
 	if err != nil {
 		return nil, RangeInfo{}, errors.Trace(err)
 	}
 
-	r, err := ParseRangeInfo(result.ContentRange)
-	if err != nil {
-		return nil, RangeInfo{}, errors.Trace(err)
+	var r RangeInfo
+	if isFullRangeRequest {
+		if result.ContentLength == nil {
+			return nil, RangeInfo{}, errors.Annotatef(berrors.ErrStorageUnknown, "open file '%s' failed. The S3 object has no content length", path)
```

Comment:
This message is not correct if open zero-size file actually returns the ContentLength. BTW, we should allow open a zero-sized file and when read from it, the result should be io.EOF not berrors.ErrStorageUnknown

Reply:
We should expect `Content-Length` in response even object size is zero. Or there is no way to get the response size.

Reply:
I've added some comments explaining that in the code.